### PR TITLE
[soltest] Test filter for isoltest

### LINKED ISSUE #6255: [soltest] Introduce test case selector for isoltest
It should be possible to selectively run certiain test cases in isoltest. For this reason we should introduce a selector for the commandline that works similarly to the boost test case selector.

### PR DESCRIPTION
**Pull request by erak**
Closes https://github.com/ethereum/solidity/issues/6255.

This PR adds the option `--test`or in short `-t` to `isoltest`. It can be used to filter the tests to be executed like this:
```
isoltest --test */* (default, needs to be quoted)
isoltest -t yulOptimizerTests/*
isoltest -t smt*/*
isoltest -t syntaxTests/array/*
isoltest -t syntaxTests/array/array_pop_arg
isoltest -t syntaxTests/array/array_*
isoltest -t */array*
```
if `--test` is not defined it defaults to `*/*` and all tests are executed. 

Invalid filter values are e.g.:
```
isoltest -t syntaxTests/
isoltest -t syntaxTests//
```

### CODE REVIEW

File: test/tools/isoltest.cpp [hunk 1]
```diff
@@ -383,7 +446,17 @@ int main(int argc, char const *argv[])
 		if (ts.smt && options.disableSMT)
 			continue;
 
-		if (auto stats = runTestSuite(ts.title, options.testPath / ts.path, ts.subpath, options.ipcPath.string(), ts.testCaseCreator, !options.noColor, options.evmVersion()))
+		auto stats = runTestSuite(
```
Comment (erak) [inline test/tools/isoltest.cpp:431]: This needs some refactoring, but that could be a separate PR.

File: test/tools/IsolTestOptions.h [hunk 2]
```diff
@@ -30,11 +30,13 @@ namespace test
...
 struct IsolTestOptions: CommonOptions
 {
-	bool noColor = false;
 	bool showHelp = false;
+	bool noColor = false;
+	std::string runTest = std::string{};
 
 	IsolTestOptions(std::string* _editor);
 	bool parse(int _argc, char const* const* _argv) override;
+	void validate() const override;
```
Comment (leonardoalt) [inline test/tools/IsolTestOptions.h:39]:
```suggestion
	void validateTest() const override;
```
Or smth like that?

Comment (erak) [reply test/tools/IsolTestOptions.h:39]: This overrides `CommonOptions::validate()` and would take care of all options, not only `--test` like in this case.

Comment (leonardoalt) [reply test/tools/IsolTestOptions.h:39]: Ok!

File: test/tools/isoltest.cpp (hunk 1 above)
Comment (erak) [reply test/tools/isoltest.cpp:431]: Partially done.

File: test/tools/IsolTestOptions.cpp [hunk 3]
```diff
@@ -51,10 +56,10 @@ IsolTestOptions::IsolTestOptions(std::string* _editor):
 	CommonOptions(description)
 {
 	options.add_options()
+		("editor", po::value<std::string>(_editor)->default_value(editorPath()), "Path to editor for opening test files.")
 		("help", po::bool_switch(&showHelp), "Show this help screen.")
-		("no-color", po::bool_switch(&noColor), "don't use colors")
-		("editor", po::value<std::string>(_editor)->default_value(editorPath()), "editor for opening test files");
-
+		("no-color", po::bool_switch(&noColor), "Don't use colors.")
+		("test,t", po::value<std::string>(&test)->default_value(std::string{}), "Filters which test units to include.");
```
Comment (erak) [inline test/tools/IsolTestOptions.cpp]: This expresses a long and a short option: `"test,t"` -> `--test | -t`

File: test/tools/isoltest.cpp [hunk 4]
```diff
@@ -60,17 +64,43 @@ struct TestStats
 	}
 };
 
+class TestFilter
```
Comment (erak) [inline test/tools/isoltest.cpp:67]: Does this need some documentation?

File: test/tools/isoltest.cpp [hunk 5]
```diff
@@ -60,17 +64,43 @@ struct TestStats
...
+
+		boost::replace_all(filter, "/", "\\/");
+		boost::replace_all(filter, "*", ".*");
+
+		m_filterExpression = regex{"(" + filter + "(\\.sol|\\.yul))"};
+	}
+
+	bool matches(string const& _name) const
+	{
+		return m_filter.empty() || regex_match(_name, m_filterExpression);
```
Comment (leonardoalt) [inline test/tools/isoltest.cpp]: Why is this done here and in `IsolTestOptions`?

File: test/tools/isoltest.cpp [hunk 6]
```diff
@@ -60,17 +64,43 @@ struct TestStats
...
 {
 public:
 	TestTool(
-		TestCase::TestCaseCreator _testCaseCreator,
-		string const& _name,
+		TestCreator _testCaseCreator,
+		TestOptions const& _options,
 		fs::path const& _path,
-		string const& _ipcPath,
-		bool _formatted,
```
Comment (leonardoalt) [inline test/tools/isoltest.cpp:71]: Why is this gone? Is it meant to be part of `IsolTestOptions`?

Comment (leonardoalt) [reply test/tools/isoltest.cpp:71]: Ah ok, it is inferred internally from `noColor`

File: test/tools/isoltest.cpp (hunk 5 above)
Comment (erak) [reply test/tools/isoltest.cpp]: `IsolTestOptions` checks, if the value of `-t` is specified and valid filter expression (e.g. `semanticTests/*` or `semanticTests/array*` etc.). This checks if the filter is not set (all tests match) or if it's set and filename matches the filter.

File: test/tools/isoltest.cpp (hunk 4 above)
Comment (ekpyron) [reply test/tools/isoltest.cpp:67]: I'd say it's fine.

File: test/tools/IsolTestOptions.cpp (hunk 3 above)
Comment (ekpyron) [reply test/tools/IsolTestOptions.cpp]: While we're add it you could also add ``-e`` for ``--editor`` and especially ``-h`` for ``--help``... but optional for this PR, but it'd be fine to just put it in, if you're still changing stuff anyways.

**Review by ekpyron — DISMISSED** [review]

**Comment by erak** [comment]
Needs rebase after https://github.com/ethereum/solidity/pull/6513 was merged.

**Comment by chriseth** [comment]
Rebased.

**Comment by leonardoalt** [comment]
`*` and `*/` run 0 tests.
`*/*` issues `Invalid test unit filter`
I guess that's fine. If you want to run everything just don't use a filter.

**Review by leonardoalt — APPROVED** [review]

**Comment by chriseth** [comment]
@leonardoalt I think the issue about `*` is due to the shell expanding it. If you want to run all tests you have to use `'*'` or a different shell :)
